Moving locals command to xplat

mishra14 opened the pull request:
Objectives - 
1. Move locals command into nuget.core.commandline.xplat
2. Add plumbing to allow dotnet nuget locals
3. Add Clear Nuget Cache button into VS package Manager options - https://github.com/NuGet/Home/issues/3289
4. Add test cases for xplat locals.

cc : @drewgil @rohit21agrawal @alpaix @joelverhagen @jainaashish @rrelyea 

Nvm the number of commits below. I'll be sure to squash. :)

Commit:
starting work on xplatting locals

Commit:
starting work on xplatting locals

Commit:
Adding Machine Cache from nuget2 core.dll and its dependencies

Commit:
Adding Machine Cache from nuget2 core.dll and its dependencies

Commit:
Removing non-needed files

Commit:
Removing non-needed files

Commit:
Prelim testing on windows exe

Commit:
Prelim testing on windows exe

Commit:
Adding package cache to see test results

Commit:
Adding package cache to see test results

Commit:
Refactoring for the PR

Commit:
Refactoring for the PR

Commit:
removing redundant file

Commit:
removing redundant file

Commit:
Adding clear cache button into VS settings

Commit:
Adding clear cache button into VS settings

Commit:
Removing package cache fromlocalc command and tests

Commit:
Removing package cache fromlocalc command and tests

Commit:
Fixing locals command description

Commit:
Fixing locals command description

Commit:
Adding Locals command in xplat

Commit:
Adding Locals command in xplat

Commit:
Refactoring strings

Commit:
Refactoring strings

Commit:
fixing flags in locals command parsing

Commit:
fixing flags in locals command parsing

Commit:
registering locals

Commit:
registering locals

Commit:
removing config file flag

Commit:
removing config file flag

Commit:
fixing confilcts

Commit:
fixing confilcts

Commit:
Adding lisence info to new files

Commit:
Adding argument check to xplat locals command

Commit:
fixing error in help locals xplat

Commit:
Fixing error message

Commit:
fixing xplat -c description

Commit:
Adding initial test cases for xplat locals

Commit:
Fixing resource name error message

Commit:
Adding tests for xplat locals

Commit:
fixing error message to match new flags

Commit:
Adding locals command runner result enum

Commit:
Adding popup forVS nuget clear cache button result

Commit:
small fixes to simplify the diff

Commit:
Fixing flags test case

alpaix commented on src/NuGet.Clients/Options/GeneralOptionControl.cs:
```diff
@@ -1,10 +1,13 @@
+using NuGet.Configuration;
+using NuGet.PackageManagement.VisualStudio;
 using System;
+using System.Collections.Generic;
```
sort and format the document. `System.*` should be on top

alpaix commented on src/NuGet.Clients/Options/GeneralOptionControl.cs:
```diff
@@ -98,5 +101,26 @@ private void packageRestoreConsentCheckBox_CheckedChanged(object sender, EventAr
+        private void localsCommandButton_OnClick(object sender, EventArgs e)
+        {
+            var arguments = new List<string>();
+            arguments.Add("all");
```
tip. Use inplace collection initialization

``` c#
var arguments = new List<string> { "all" }
```

alpaix commented on src/NuGet.Clients/Options/GeneralOptionControl.cs:
```diff
@@ -98,5 +101,26 @@ private void packageRestoreConsentCheckBox_CheckedChanged(object sender, EventAr
+            {
+                MessageBox.Show(ex.Message, Resources.PopupTitle_ClearNuGetCache);
+            }
+            if (localsCommandRunner.Result == LocalsCommandRunner.LocalsCommandResult.ClearSuccess)
```
How do we notify notify user on error? Popup, ActivityLog, Telemetry?

mishra14 replied:
A message box/popup. Same as below, but with the exception message (Which will be the same as the message displayed for cli errors)

Ref - 

![2016-08-17](https://cloud.githubusercontent.com/assets/10507120/17746904/34d983f6-6467-11e6-8092-682dae58eea9.png)

alpaix replied:
Then you should handle `InvalidLocalResourceName` and `ClearFailure` results.

jainaashish replied:
OK button on dialog not aligned

jainaashish replied:
More over dialog box title should reflect whether operation failed or succeed

alpaix commented on src/NuGet.Clients/Options/GeneralOptionControl.cs:
```diff
@@ -98,5 +101,26 @@ private void packageRestoreConsentCheckBox_CheckedChanged(object sender, EventAr
+            {
+                localsCommandRunner.ExecuteCommand();
+            }
+            catch(ArgumentException ex)
```
What about other exception types?

mishra14 replied:
Sure. Will change ArgumentException to Exception. Will catch all exceptions.

alpaix replied:
Also log error to the ActivityLog

``` c#
ActivityLog.LogError(NuGetUI.LogEntrySource, ex.ToString());
```

jainaashish replied:
nit: formatting

alpaix commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/LocalsCommand.cs:
```diff
@@ -0,0 +1,64 @@
+using NuGet.Commands;
+using NuGet.Common;
+using NuGet.Configuration;
+using System;
```
`System.*` on top

alpaix commented on src/NuGet.Core/NuGet.Commands/LocalsCommandRunner.cs:
```diff
@@ -0,0 +1,238 @@
+
+using NuGet.Common;
+using NuGet.Configuration;
+using System;
```
same here

alpaix commented on src/NuGet.Core/NuGet.Commands/LocalsCommandRunner.cs:
```diff
@@ -0,0 +1,238 @@
+
+namespace NuGet.Commands
+{
+    public class LocalsCommandRunner
```
Document class and public methods

alpaix commented on src/NuGet.Core/NuGet.Commands/LocalsCommandRunner.cs:
```diff
@@ -0,0 +1,238 @@
+            ClearFailure,
+            ClearSuccess
+        }
+        private const string _httpCacheResourceName = "http-cache";
```
constants should be "Pascal" case `HttpCacheResourceName`

mishra14 replied:
Picked this up from existing code in Nuget.clients.Localscommand.cs and refactored it as is.
Should I alter this then?

mishra14 replied:
Nvm. Will fix this.

alpaix replied:
The point is to use this opportunity to review the refactored code and make necessary changes.

alpaix commented on src/NuGet.Core/NuGet.Commands/LocalsCommandRunner.cs:
```diff
@@ -0,0 +1,238 @@
+
+        public LocalsCommandRunner(IList<string> arguments, ISettings settings, bool clear, bool list)
+        {
+            this.Arguments = arguments;
```
avoid using `this`

alpaix replied:
null check

alpaix commented on src/NuGet.Core/NuGet.Commands/LocalsCommandRunner.cs:
```diff
@@ -0,0 +1,238 @@
+
+            if (failedDeletes.Any())
+            {
+                Console.WriteLine(String.Format(CultureInfo.CurrentCulture, Strings.LocalsCommand_LocalsPartiallyCleared));
```
This message should be logged via logger

alpaix commented on src/NuGet.Core/NuGet.Commands/LocalsCommandRunner.cs:
```diff
@@ -0,0 +1,238 @@
+
+                foreach (var failedDelete in failedDeletes.OrderBy(f => f, StringComparer.OrdinalIgnoreCase))
+                {
+                    Console.WriteLine(String.Format(CultureInfo.CurrentCulture, Strings.LocalsCommand_FailedToDeletePath, failedDelete));
```
log error

alpaix commented on src/NuGet.Core/NuGet.Common/PathUtil/NuGetEnvironment.cs:
```diff
@@ -256,7 +256,7 @@ public static string GetDotNetLocation()
         /// built-in functionality.
         /// </summary>
-        private enum SpecialFolder
+        public enum SpecialFolder
```
Should be in a separate file once it's public

jainaashish replied:
why you need to make it public?

alpaix commented on src/NuGet.Core/NuGet.Common/MachineCache.cs:
```diff
@@ -0,0 +1,56 @@
+#endif
+        }
+
+        public static string GetCachePath(Func<string, string> getEnvironmentVariable, string localAppDataPath)
```
private?

alpaix commented on src/NuGet.Core/NuGet.Common/MachineCache.cs:
```diff
@@ -0,0 +1,56 @@
+    /// The machine cache represents a location on the machine where packages are cached. It is a specific implementation of a local repository and can be used as such.
+    /// NOTE: this is a shared location, and as such all IO operations need to be properly serialized
+    /// </summary>
+    public class MachineCache
```
`public static class` as it doesn't contain instance methods

alpaix commented on src/NuGet.Core/NuGet.Common/MachineCache.cs:
```diff
@@ -0,0 +1,56 @@
+        public static string GetCachePath()
+        {
+#if IS_CORECLR
+            string localAppDataPath = NuGetEnvironment.GetFolderPath(NuGetFolderPath.HttpCacheDirectory);
```
consider initializing `localAppDataPath` and `getEnvironmentVariable` in a static constructor

alpaix commented on src/NuGet.Core/NuGet.Common/MachineCache.cs:
```diff
@@ -0,0 +1,56 @@
+    /// </summary>
+    public class MachineCache
+    {
+        private const string NuGetCachePathEnvironmentVariable = "NuGetCachePath";
```
We used to have environment variables in all-capital case like `NUGET_SHOW_STACK`. Is this an exception?

mishra14 replied:
Again, refactored from old MachineCache.cs 

Should I change this as well?

alpaix replied:
Is this existing and documented variable?

alpaix commented on src/NuGet.Core/NuGet.Common/MachineCache.cs:
```diff
@@ -0,0 +1,56 @@
+
+        public static string GetCachePath(Func<string, string> getEnvironmentVariable, string localAppDataPath)
+        {
+            string cacheOverride = getEnvironmentVariable(NuGetCachePathEnvironmentVariable);
```
`var`

alpaix commented on src/NuGet.Core/NuGet.Common/MachineCache.cs:
```diff
@@ -0,0 +1,56 @@
+            string localAppDataPath = NuGetEnvironment.GetFolderPath(NuGetFolderPath.HttpCacheDirectory);
+            return GetCachePath(Environment.GetEnvironmentVariable, localAppDataPath);
+#else
+            string localAppDataPath = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
```
`var`

alpaix commented on src/NuGet.Core/NuGet.Common/MachineCache.cs:
```diff
@@ -0,0 +1,56 @@
+        public static string GetCachePath(Func<string, string> getEnvironmentVariable, string localAppDataPath)
+        {
+            string cacheOverride = getEnvironmentVariable(NuGetCachePathEnvironmentVariable);
+            if (!String.IsNullOrEmpty(cacheOverride))
```
[type name](https://github.com/NuGet/Home/wiki/Coding-Guidelines#use-c-type-keywords-in-favor-of-net-type-names)

alpaix commented on src/NuGet.Core/NuGet.Common/MachineCache.cs:
```diff
@@ -0,0 +1,56 @@
+                {
+                    // there's a bug on Windows Azure Web Sites environment where calling through the Environment.GetFolderPath()
+                    // will returns empty string, but the environment variable will return the correct value
+                    localAppDataPath = getEnvironmentVariable("LocalAppData");
```
Introduce a constant for string values

alpaix commented on src/NuGet.Core/NuGet.Common/MachineCache.cs:
```diff
@@ -0,0 +1,56 @@
+
+                if (String.IsNullOrEmpty(localAppDataPath))
+                {
+                    return null;
```
Is the caller aware this method may return `null`?

alpaix commented on src/NuGet.Core/NuGet.Common/MachineCache.cs:
```diff
@@ -0,0 +1,56 @@
+        /// <summary>
+        /// Determines the cache path to use for NuGet.exe. By default, NuGet caches files under %LocalAppData%\NuGet\Cache.
+        /// This path can be overridden by specifying a value in the NuGetCachePath environment variable.
+        /// </summary>
```
Add `<returns>...</returns>` and specify the method may return `null`

alpaix commented:
:watch:

jainaashish commented on src/NuGet.Clients/NuGet.CommandLine/Commands/LocalsCommand.cs:
```diff
@@ -1,12 +1,9 @@
-using System.Collections.Generic;
+using NuGet.Commands;
 using System.Linq;
 using System.Threading.Tasks;
```
correct the order of usings

jainaashish commented on src/NuGet.Clients/Options/GeneralOptionControl.cs:
```diff
@@ -98,5 +101,26 @@ private void packageRestoreConsentCheckBox_CheckedChanged(object sender, EventAr
+            }
+            catch(ArgumentException ex)
+            {
+                MessageBox.Show(ex.Message, Resources.PopupTitle_ClearNuGetCache);
```
Would user be able to understand these exception messages? is there any spec for this change? Also any design spec for dialog box?

mishra14 replied:
Spec : [link](https://github.com/NuGet/Home/wiki/Add-Clear-Cache-button-to-VS-Package-Manager-options)

Had a talk with hari about this. Will be making changes to the spec as well.

jainaashish commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/LocalsCommand.cs:
```diff
@@ -0,0 +1,64 @@
+                {
+                    var setting = Settings.LoadDefaultSettings(root: null, configFileName: null, machineWideSettings: null);
+                    if (((arguments.Values.Count < 1) || string.IsNullOrWhiteSpace(arguments.Values[0])) || (clear.HasValue() && list.HasValue()) || (!clear.HasValue() && !list.HasValue()))
+                    {
```
format the doc, break long lines.

jainaashish commented on src/NuGet.Core/NuGet.Commands/LocalsCommandRunner.cs:
```diff
@@ -0,0 +1,238 @@
+        /// <returns></returns>
+        public void ExecuteCommand()
+        {
+            var localResourceName = GetLocalResourceName(Arguments[0]);
```
check if Arguments list empty...

mishra14 replied:
Checks are done by callers - 
[LocalsCommand.cs](https://github.com/NuGet/NuGet.Client/blob/ecac235f6541a7673e544a157bb7b7d64cb91338/src/NuGet.Clients/NuGet.CommandLine/Commands/LocalsCommand.cs#L25)
[Xplat Locals Command](https://github.com/NuGet/NuGet.Client/blob/ecac235f6541a7673e544a157bb7b7d64cb91338/src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/LocalsCommand.cs#L44)

Since both callers have a different way of handling errors, I have not consolidated this.

jainaashish replied:
And why do we want to handle it separately? Also i think good practice is to handle such things in called func like what happens if tomorrow someone else also calls this func and pass an empty list

mishra14 replied:
Drew is working on migrating help into Nuget.core. Once that is done, I can use that for consolidating the two into LocalsCommandRunner. What do you think ?

jainaashish commented on test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/XPlatLocalsTests.cs:
```diff
@@ -0,0 +1,63 @@
+
+namespace NuGet.XPlat.FuncTest
+{
+    public class XPlatLocalsTests
```
Could add few more tests like:
- locals all -l
- locals all --list
- locals --clear all
- locals -l < specific source >

Commit:
Fixing based on PR comments, adding logging to localscommandrunner and fixing error handling in options control UI

Commit:
removing non-needed changes and logs

Commit:
removing non-needed changes and logs

Commit:
Improving UI error message

mishra14 commented:
@jainaashish @alpaix : Updated

Changes - 
1. Addressed most of the PR comments
2. Adding Logging into LocalsCommandRunner - Had to move couple of files into Nuget.Console from Nuget.VSExtension
3. Changed the UI on Clear cache button press within VS settings, based on feedback from Hari. Will have a spec review once and this might change. Ref - [updated spec](https://github.com/NuGet/Home/wiki/Add-Clear-Cache-button-to-VS-Package-Manager-options)

alpaix commented on src/NuGet.Clients/Options/GeneralOptionControl.cs:
```diff
@@ -2,22 +2,33 @@
     {
         private readonly Configuration.ISettings _settings;
         private bool _initialized;
+        private IServiceProvider _serviceprovider;
```
readonly

alpaix commented on src/NuGet.Clients/Options/GeneralOptionControl.cs:
```diff
@@ -2,22 +2,33 @@
         private readonly Configuration.ISettings _settings;
         private bool _initialized;
+        private IServiceProvider _serviceprovider;
+        private OutputConsoleLogger _outputConsoleLogger;
```
readonly

alpaix commented on src/NuGet.Clients/Options/GeneralOptionControl.cs:
```diff
@@ -2,22 +2,33 @@
+        private OutputConsoleLogger _outputConsoleLogger;
 
-        public GeneralOptionControl()
+        public GeneralOptionControl(IServiceProvider serviceProvider)
```
null check

alpaix commented on src/NuGet.Clients/Options/GeneralOptionControl.cs:
```diff
@@ -98,5 +109,48 @@ private void packageRestoreConsentCheckBox_CheckedChanged(object sender, EventAr
+            updateLocalsCommandStatusText(string.Format(Resources.ShowMessage_LocalsCommandWorking), visibility: true);
+            var arguments = new List<string> { "all" };
+            var settings = ServiceLocator.GetInstance<ISettings>();            
+            Log logError = new Log(LogError);
```
var

alpaix commented on src/NuGet.Clients/Options/GeneralOptionControl.cs:
```diff
@@ -98,5 +109,48 @@ private void packageRestoreConsentCheckBox_CheckedChanged(object sender, EventAr
+            var arguments = new List<string> { "all" };
+            var settings = ServiceLocator.GetInstance<ISettings>();            
+            Log logError = new Log(LogError);
+            Log logInformation = new Log(LogInformation);
```
var

alpaix commented on src/NuGet.Clients/Options/GeneralOptionControl.cs:
```diff
@@ -98,5 +109,48 @@ private void packageRestoreConsentCheckBox_CheckedChanged(object sender, EventAr
+                LogError(string.Format(Resources.ShowMessage_LocalsCommandFailure, DateTime.Now.ToString(Resources.Culture), ex.Message));
+                ActivityLog.LogError(NuGetUI.LogEntrySource, ex.ToString());
+            }
+            if (localsCommandRunner.Result == LocalsCommandResult.ClearSuccess)
```
handle the rest of `LocalsCommandResult` values

alpaix commented on src/NuGet.Clients/VsConsole/Console/OutputConsole/Guids.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace NuGetConsole
+{
+    internal static class GuidList
```
There's another one `VsConsole/Console/Guids.cs`. Take what's really needed and copy there.

Commit:
Fixing accessors and removing redundant code

Commit:
Fixing cache type

alpaix commented:
Looks good! :clap: :shipit:

Commit:
Converting error message label to rich text to handle hyperlink within

Commit:
fixing cache type

mishra14 commented:
Merged : https://github.com/NuGet/NuGet.Client/commit/6f3ae2d59140f5ea97eb7573535de1c286d6d336
